### PR TITLE
chore: Merge Slurm-infrastructure support changes

### PR DESCRIPTION
**Pull request by jerryharrow**
## Description

One of many small changes coming to make the diff more manageable.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/launch/deepspeed.py
Comment (stoksc) [inline harness/determined/launch/deepspeed.py:7]: I think everything in this file is good, already (so we can revert this)

File: master/pkg/archive/archive.go
Comment (stoksc) [inline master/pkg/archive/archive.go:97]: just for consistency, can you make `RootItem` above behave the same? or, maybe better, remove "Linkname" from "Item" and do the mapping with the rest of the logic in "func tarArchive"

File: harness/determined/launch/horovod.py
Comment (stoksc) [inline harness/determined/launch/horovod.py:22]: @rb-determined-ai can you just look at this file instead of my pr. @jerryharrow wrote it so I'll let him pr it.

File: master/pkg/tasks/task.go
Comment (stoksc) [inline master/pkg/tasks/task.go:133]:
```suggestion
		e["GLOO_PORT_RANGE"] = t.TaskContainerDefaults.GlooPortRange
```

File: master/pkg/tasks/task.go
Comment (stoksc) [inline master/pkg/tasks/task.go:156]:
```suggestion
			// TODO(DET-7565): Propagate errors.
```

Comment (stoksc) [reply master/pkg/tasks/task.go:156]: made a ticket

File: master/pkg/tasks/task.go
Comment (stoksc) [inline master/pkg/tasks/task.go:293]: maybe @rb-determined-ai should review these lines, since he wrote our libnss stuff.

File: master/static/srv/entrypoint.sh
Comment (stoksc) [inline master/static/srv/entrypoint.sh:6]: maybe drop a comment explaining that slurm sets `CUDA_VISIBLE_DEVICES`

Comment (stoksc) [reply master/static/srv/entrypoint.sh:6]: also, we probably need this in every entrypoint, so maybe pull it into a script and use that?

File: master/static/srv/shell-entrypoint.sh
Comment (stoksc) [inline master/static/srv/shell-entrypoint.sh:49]:
```suggestion
# of the form, and a weird %% definition.
```
and some other spelling/grammar stuff in the comment

File: master/static/srv/task-logging-setup.sh
Comment (stoksc) [inline master/static/srv/task-logging-setup.sh:75]:
```suggestion
    exec 1> >(tee -p >("$DET_PYTHON_EXECUTABLE" /run/determined/enrich_task_logs.py --stdtype stdout) >&1; printf x  >$DET_LOG_WAIT_FIFO) \
```
here and on the other, since how we did this changed out from under us.

**Review by stoksc — COMMENTED** [review]
seems like `enrich_task_logs.py` didn't actually make it here

File: harness/determined/launch/deepspeed.py
Comment (jerryharrow) [reply harness/determined/launch/deepspeed.py:7]: I did figure that out earlier and attempted to revert, but apparently not :-)

File: master/pkg/tasks/task.go
Comment (jerryharrow) [reply master/pkg/tasks/task.go:133]: Of course there is no such thing, which makes that a non-trivial exercise.   
I

**Comment by jerryharrow** [comment]
I dropped the code that has to be re-worked from this review as fixes need to be made/tested on the dispatcher branch before I can move them here.

**Comment by jerryharrow** [comment]
> seems like `enrich_task_logs.py` didn't actually make it here

The Pods unit tests were failing now that we added enrich_task_logs.py into the helper archive (and it doesn't exist in this tree).   I was just creating a stub for that, but I realize now what is happening.    The same test was failing in the same way in the dispatcher tree, but today it started succeeding again.   I will bring the full version of enrich_task_logs.py  and drop in in the appropriate place here -- unless we should instead drop it from the run archive (and put it in via the dispatcher-task.go configuration -- like other slurm-specific files)?

File: master/static/srv/entrypoint.sh
Comment (jerryharrow) [reply master/static/srv/entrypoint.sh]: Addressing on dispatcher branch.

File: master/pkg/tasks/task.go
Comment (jerryharrow) [reply master/pkg/tasks/task.go]: Updated the comment.

File: master/pkg/archive/archive.go
Comment (jerryharrow) [reply master/pkg/archive/archive.go]: Will address on dispatcher branch where I can test it, before merging.

File: master/static/srv/task-logging-setup.sh
Comment (stoksc) [reply master/static/srv/task-logging-setup.sh]: I think you can just fix here and I can fix on rebase

File: master/pkg/tasks/task.go
Comment (stoksc) [reply master/pkg/tasks/task.go:133]: should be? https://github.com/determined-ai/determined/blob/7b3fee44c0846f3546f7e4bf57b231d227dcd27d/master/pkg/model/task_container_defaults.go#L23

Comment (jerryharrow) [reply master/pkg/tasks/task.go:133]: Ohh.  It is GLOOPortRange, not GlooPortRange -- grep didn't find what you suggested.

**Comment by stoksc** [comment]
> unless we should instead drop it from the run archive (and put it in via the dispatcher-task.go configuration -- like other slurm-specific files)?

this is probably better, but I think either is fine.

File: harness/determined/launch/horovod.py
Comment (rb-determined-ai) [inline harness/determined/launch/horovod.py:113]: stray print

File: harness/determined/launch/horovod.py
Comment (rb-determined-ai) [inline harness/determined/launch/horovod.py:202]: Maybe just a plain:
```python
os.environ.pop("SLURM_JOBID", None)
```
and we can make a loop out of it when we need it to be a loop.

File: master/pkg/tasks/task.go
Comment (rb-determined-ai) [reply master/pkg/tasks/task.go:293]: Cat we use `*` instead of `BogusPasswordHash`?

I see from `man 5 shadow` that `*` or `!` would both cause password logins to fail, though `!` has the special meaning that the account is locked (which I suppose something in the slurm/singularity stack is looking at).

Comment (jerryharrow) [reply master/pkg/tasks/task.go:293]: Right.  !  was completely blocked.   I can try * instead if you prefer, but gonna delay this til next week, since I can't test on this branch.   I can only test this support on the determined-ee dispatcher branch at the moment.

Comment (jerryharrow) [reply master/pkg/tasks/task.go]: It wasn't actually that hard to test -- so pushed up the change to use * instead.

**Review by stoksc — APPROVED** [review]
go side lgtm

**Comment by stoksc** [comment]
I'll wait on Ryan for his bits

File: master/pkg/tasks/task.go
Comment (rb-determined-ai) [inline master/pkg/tasks/task.go:292]: is this line still necessary?  Or is it from the old version?

I don't understand why PasswordAuthentication makes a difference as to whether or not `*` works.

File: master/static/srv/enrich_task_logs.py
Comment (rb-determined-ai) [inline master/static/srv/enrich_task_logs.py:32]: A 5-second buffer seems pretty high.  Certainly watching trial logs would be a jarring experience.  Is .5 sec too short?

File: master/static/srv/enrich_task_logs.py
Comment (rb-determined-ai) [inline master/static/srv/enrich_task_logs.py:34]: This is always going to block, even long after the 5-second buffer timeout has passed.

You should either use selectors, or use two threads with a queue in between, where the queue has a timeout while waiting for new lines.

File: master/static/srv/enrich_task_logs.py
Comment (jerryharrow) [reply master/static/srv/enrich_task_logs.py:32]: I'm just merging this code, it is only used by Slurm.    You should take it up with @stoksc because he directed the implementation :-)

File: master/pkg/tasks/task.go
Comment (jerryharrow) [reply master/pkg/tasks/task.go:292]: Updated the comment.

File: master/static/srv/enrich_task_logs.py
Comment (jerryharrow) [reply master/static/srv/enrich_task_logs.py:34]: For @stoksc  as well

File: master/static/srv/enrich_task_logs.py
Comment (stoksc) [reply master/static/srv/enrich_task_logs.py:32]: haha. I mean, for what it is worth I nearly insisted we work off https://github.com/determined-ai/determined-ee/pull/185/files but we took this approach instead.

Comment (stoksc) [reply master/static/srv/enrich_task_logs.py:32]: @rb-determined-ai can we fly with this for now, I'll try to land my batching over the next day?

Comment (rb-determined-ai) [reply master/static/srv/enrich_task_logs.py:32]: sure I guess.  Still working on one more view comment though.

File: master/static/srv/shell-entrypoint.sh
Comment (rb-determined-ai) [inline master/static/srv/shell-entrypoint.sh:59]: I'm thinking that we've hit the limit of this blacklist/sed strategy for parsing environment variables.  For example, we are acknowledging that we are dealing with multiline variables, but the grep command is assuming the multiline variables never match `^\S+=\S`.

How about we use `env -0` to get NUL-byte delimited environment variables, then use `mapfile -d ''` to read them into a bash array, then we iterate through the bash array:

```bash
# extglob enables +() notation in patterns of ${parameter/pattern/string} notation
shopt -s extglob

# convert NUL-delimited environment key-value pairs in an array
# -d '' means "use NUL byte as delimiter"
# -t means "strip the delimiter"
mapfile -d '' -t kvps < <(env -0)

# iterate through each key-value pair in the array
options="$(
    for kvp in "${kvps[@]}" ; do
        # Variable name is what comes before the first '='
        var="${kvp/=*/}"
        # Variable content starts after the first '='
        val="${kvp/#+([^=])=/}"

        # Filter names we shouldn't forward
        if [[ "$var" =~ ^(_|HOME|TERM|LANG|LC_.*)$ ]] ; then
            continue
        fi

        # For slurm: filter variables with %% or () in the name
        if [[ "$var" =~ (%%|\(\)) ]] ; then
            continue
        fi

        # Backslash-escape quotes so that sshd works.
        val="${val//\"/\\\"}"
        # Backslash-escape backslashes so that sed doesn't interpret them.
        val="${val//\\/\\\\}"
        # Backslash-escape forward slashes so that sed works.
        val="${val////\\/}"
        echo -n "environment=\"$var=$val\","
    done
)"
```

I tested with:
```
env -i \
 LC_ALL=1 \
 ASDF=A=B=C \
 "$(echo -e 'newline=before\nafter')" \
 "PERCENTS%%=ZXCV" \
 "PARENS()=ZXCV" \
 bash test.sh
```

to confirm that equals and newlines in variables are preserved, but LC_* is still ignored, as are %% and () if they appear.

File: master/static/srv/enrich_task_logs.py
Comment (stoksc) [reply master/static/srv/enrich_task_logs.py:32]: @rb-determined-ai https://github.com/determined-ai/determined-ee/pull/305/files I'll bring it back over once this lands

File: master/static/srv/shell-entrypoint.sh
Comment (jerryharrow) [reply master/static/srv/shell-entrypoint.sh]: Done

File: master/pkg/tasks/task.go
Comment (rb-determined-ai) [inline master/pkg/tasks/task.go:131]: can we just drop this?

We once had a fork of NCCL that supported setting port ranges, but we no longer maintain that fork or ship it to users

It's basically an ignored config field now.

(the GLOO_PORT_RANGE is built into gloo, so it's still valid).

Comment (jerryharrow) [reply master/pkg/tasks/task.go:131]: This allows users to override the GLOO port range if they are hitting problems.
With Slurm+Singularity the ports are global on the compute nodes for all experiments.   If more than one experiment is running on the same node there can be conflicts.

Comment (jerryharrow) [reply master/pkg/tasks/task.go:131]: So I'm not sure what's up here.      

chore: remove legacy code/docs for NCCL/Gloo port range config (#4187)

Which removed it, but @stoksc said earlier in this review to change it from:

`e["GLOO_PORT_RANGE"] = t.TaskContainerDefaults.NCCLPortRange`

to 

`e["GLOO_PORT_RANGE"] = t.TaskContainerDefaults.GLOOPortRange`

I guess I can just drop it, unless  @stoksc was really trying to put it back.

Comment (stoksc) [reply master/pkg/tasks/task.go:131]: Let's just leave it for now and I can PR to remove the config entirely from the master shortly?

Comment (rb-determined-ai) [reply master/pkg/tasks/task.go:131]: I think removing it from the config is technically a breaking feature, which is why it was still allowed in the config even after #4187.  

But `NCCL_PORT_RANGE` is an environment variable that [I personally added to our nccl fork three years ago](https://github.com/determined-ai/nccl/commit/8768cb3f881ce198825294b62a435e3f3a804baf).  That fork is no longer used anywhere, and upstream nccl doesn't use `NCCL_PORT_RANGE`, so setting `NCCL_PORT_RANGE` will have no effect (except maybe confusing a user somewhere).

I'm not talking about removing the GLOO_PORT_RANGE environment variable, only the NCCL_PORT_RANGE.

**Review by rb-determined-ai — APPROVED** [review]
great work!

File: master/static/srv/enrich_task_logs.py
Comment (stoksc) [reply master/static/srv/enrich_task_logs.py]: pls and ty

File: master/static/srv/enrich_task_logs.py
Comment (rb-determined-ai) [inline master/static/srv/enrich_task_logs.py:62]: it occurs to me, the slurm logs will have the extra newlines.  Can you put `end=""` in here?

And, can you maybe leave a comment explaining that we are intentionally preserving stdin on stderr so that slurm collects the logs, in addition to shipping them to determined?

Comment (jerryharrow) [reply master/static/srv/enrich_task_logs.py:62]: Correct.   Will fix.

**Review by rb-determined-ai — APPROVED** [review]
looks good

**Pull request merged**
